### ISSUE TITLE
[build-system][MSBuild]Build error after upgrade

### ISSUE THREAD

**Issue by d0nc1h0t** [conan-io/conan#5596]
Conan version 1.18.1
Windows 10 Pro
Visual Studio 2017 Community Edition (version 15) / Visual Studio 2019 Community Edition (version 16)
python 2.7.16
pip 19.2.1

After upgrading to version newer 1.15.4, the MSBuild system gives an error during building.

For reproduce:

git clone https://github.com/odant/conan-tcmalloc .
conan create . test/testing

**Comment by jgsogo** [comment]
I'm not running Windows right now... can you paste the error here, please?

**Comment by d0nc1h0t** [comment]
Sounds like a path problem: 

> Cannot open include file: 'config.h': No such file or directory

`       ClCompile:
         C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\bin\HostX64\x64\CL.exe /c /ZI /JMC /nologo /W3 /WX- /diagnostics:column /Od /D _DEBUG /D _WINDLL /D _UNICODE /D UNICODE /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Zc:rvalueCast /Fo"x64\Debug\\" /Fd"x64\Debug\vc142.pdb" /Gd /TC /FC /errorReport:queue ..\..\src\base\dynamic_annotations.c
         dynamic_annotations.c
     2>D:\.conan\data\tcmalloc\2.9.111.222\odant\testing\build\f7b34870a1ff2ae4b6b42738dd0293766e41e6cf\src\src\base\dynamic_annotations.c(38,10): error C1083:  Cannot open include file: 'config.h': No such file or directory`

At the end, an error is displayed:

`ERROR: tcmalloc/2.9.111.222@odant/testing: Error in build() method, line 74
        self.msvc_build()
while calling 'msvc_build', line 86
        builder.build("gperftools.sln", upgrade_project=False, verbosity="normal", use_env=False, build_type=build_type, targets=["libtcmalloc_minimal"])
        ConanException: Error 1 while executing echo Conan:vcvars already set && msbuild "gperftools.sln" /p:Configuration="Debug" /p:Platform="x64" /m:12 /target:libtcmalloc_minimal /p:PlatformToolset="v142" /verbosity:normal /p:ForceImportBeforeCppTargets="D:\.conan\data\tcmalloc\2.9.111.222\odant\testing\build\f7b34870a1ff2ae4b6b42738dd0293766e41e6cf\src\conan_build.props"`

When executing commands from the error text manually (change path and just copy-past command), the building process is completed successfully.

`cd D:\.conan\data\tcmalloc\2.9.111.222\odant\testing\build\f7b34870a1ff2ae4b6b42738dd0293766e41e6cf\src
msbuild "gperftools.sln" /p:Configuration="Debug" /p:Platform="x64" /m:12 /target:libtcmalloc_minimal /p:PlatformToolset="v142" /verbosity:normal /p:ForceImportBeforeCppTargets="D:\.conan\data\tcmalloc\2.9.111.222\odant\testing\build\f7b34870a1ff2ae4b6b42738dd0293766e41e6cf\src\conan_build.props"`

**Comment by memsharded** [comment]
Confirmed this changed between 1.15.4 and 1.16.0. To reproduce I needed to remove the ``def build_requirements(self):`` in Windows.

**Comment by memsharded** [comment]
The change affecting this is:

```diff
(conanpip27) λ git diff release/1.15.4 -- conans\client\build\visual_environment.py
diff --git a/conans/client/build/visual_environment.py b/conans/client/build/visual_environment.py
index d50eb085..37dfb9ce 100644
--- a/conans/client/build/visual_environment.py
+++ b/conans/client/build/visual_environment.py
@@ -11,6 +11,7 @@ class VisualStudioBuildEnvironment(object):
     - LIB: library paths with semicolon separator
     - CL: /I (include paths)
     - _LINK_: linker options and libraries
+    - UseEnv: True https://github.com/conan-io/conan/pull/4583

         return {"CL": cl_args,
                 "LIB": lib_paths,
-                "_LINK_": link_args}
+                "_LINK_": link_args,
+                "UseEnv": "True"}

     @property
     def vars_dict(self):
@@ -117,7 +120,8 @@ class VisualStudioBuildEnvironment(object):

         ret = {"CL": cl,
                "LIB": lib,
-               "_LINK_": link}
+               "_LINK_": link,
+               "UseEnv": "True"}
         return ret

     def _std_cpp(self):
```

Removing ``"UseEnv": "True"`` makes this pass.

**Comment by memsharded** [comment]
That UseEnv=True was recommended by @jmoutray in https://github.com/conan-io/conan/pull/4583#issuecomment-469450081, PR: https://github.com/conan-io/conan/pull/4583

**Comment by d0nc1h0t** [comment]
> To reproduce I needed to remove the `def build_requirements(self):` in Windows.

I'm sorry, I forgot about this moment. This package is not used during building, but only signs the result.

UseEnv control is already present #2830 and set to False for this project. But this patch forcibly sets the value to True and an error is reproduced #2829.

**Comment by memsharded** [comment]
It seems that when the UseEnv=True is used, the values of the environment are not overriden, but the project configuration values are not used? It is not clear from the docs: https://docs.microsoft.com/en-us/cpp/build/reference/msbuild-visual-cpp-overview?view=vs-2017

**Comment by d0nc1h0t** [comment]
> It seems that when the UseEnv=True is used, the values of the environment are not overriden, but the project configuration values are not used?

I think yes, and it confirms practical experience.

**Comment by jgsogo** [comment]
From `devenv /?` (not the same as `MSBuild`, but probably the same intention) it is clear that this flag is used to switch between environment variables and project ones:

> /useenv Use PATH, INCLUDE, LIBPATH, and LIB environment variables **instead of IDE paths** for VC++ builds.

**Comment by jgsogo** [comment]
I'm proposing a different approach to fix this issue (here https://github.com/conan-io/conan/pull/5609). I think it is not a regression but a feature not completely implemented in previous PR. 

Conan, as a tool trying to maximize reproducibility, needs to control as much as possible the environment, so using `useEnv=True` by default I think it is the way to go, nevertheless we should respect user input. Current implementation with `MSBuild.build(..., use_env=False,...)` doesn't override the default behavior introduced by the changes in `VisualStudioBuildEnvironment` (`useEnv=True`), with my PR the behavior is overridden as the command line takes precedence over the environment variables.

With this change, the example provided by @d0nc1h0t works 🎉

**Comment by jgsogo** [comment]
Here I come with more information about this issue. The `UseEnv` flag affects some **variables for compilation** (read about it related to `devenv` [here](https://docs.microsoft.com/es-es/visualstudio/ide/reference/useenv-devenv-exe?view=vs-2019)):

> Remarks: This switch affects the Visual Studio IDE in the project properties for **VC++ Directories**

So, it talks about _VC++ Directories_ and not about _Include directories_. That said, if I open VS as usual, I get this values for the **VC++ Directories** (and it successfully compiles):

![image](https://user-images.githubusercontent.com/1406456/62870577-2b041d00-bd1a-11e9-9ae5-c8a9e594736e.png)


but if I open VS using `devenv /useEnv <solutionfile.sln` I get a different thing (and the include failure for `config.h`). This shows the resulting values opening VS from a _VS native x64 console_, **notice that the name now is _VC++ Environment Variables**:

![image](https://user-images.githubusercontent.com/1406456/62870527-1c1d6a80-bd1a-11e9-9b80-1a094878595f.png)


So, the `useEnv` is about these variables that are related to compilation tools, and not to the directories we are used to. And it looks like `/useEnv` completely overrides these directories (not appending, not prepending, not adding, it just give a different set of values).

----

@d0nc1h0t , probably you should add the directories `../../src/windows` and `../../src` that are internal to your project as _Include directories_ under the _C/C++ > Directories > AdditionalIncludeDirectories_ property, wdyt?

----

PD.- I think that the fix in my PR still is needed.

**Comment by d0nc1h0t** [comment]
> probably you should add the directories `../../src/windows` and `../../src` that are internal to your project as _Include directories_ under the _C/C++ > Directories > AdditionalIncludeDirectories_ property, wdyt?

I agree, the immediate settings of the project are placed in the environment part. But the contents of the src folder is unchanged [Google Performance Tools (gperftools)](https://github.com/gperftools/gperftools) sources. If necessary, I make changes in patches at the building stage and try to do it minimally.

In general, it is clear and thanks for the detailed analysis.

**Comment by lasote** [comment]
Will be released at 1.18.2
